velox/functions/sparksql/aggregates/DecimalSumAggregate.h

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include "velox/exec/SimpleAggregateAdapter.h"
+#include "velox/type/DecimalUtil.h"
+
+namespace facebook::velox::functions::aggregate::sparksql {
+
+/// TInputType refer to the raw input data type. TSumType refer to the type of
+/// sum in the output of partial aggregation or the final output type of final
+/// aggregation.
+template <typename TInputType, typename TSumType>
+class DecimalSumAggregate {
+ public:
+  using InputType = Row<TInputType>;
+
+  using IntermediateType =
+      Row</*sum*/ TSumType,
+          /*isEmpty*/ bool>;
+
+  using OutputType = TSumType;
+
+  static constexpr bool default_null_behavior_ = false;
```

Comment:
The main reason is that Spark's decimal sum doesn't have the concept of a null group; each group is initialized with an initial value, where sum = 0 and isEmpty = true. Therefore, to maintain consistency, I need to use the parameter `nonNullGroup` in `writeIntermediateResult` (this parameter is only available when non-default-null behavior is enabled), to output a null group as sum = 0, isEmpty = true.

Reply:
I'm not sure I understand. What is "null group"? Are you trying to match intermediate results to Spark's? If so, does this matter only when the query using companion functions? Why intermediate result for all null inputs cannot be sum = NULL ? Do you need isEmpty = true to allow Spark to distinguish between all-NULL inputs and decimal overflow?

Reply:
> What is "null group"?

"null group" means all input values for this group are null, we never call clearNull for this group.

> Are you trying to match intermediate results to Spark's?

Yes. Not only for companion functions. In Spark decimal sum agg, sum is initialized to 0, not null. sum=0, isEmpty=true means all input values are null. Spark use isEmpty=true to distinguish between all-NULL inputs and decimal overflow.

Reply:
We need to ensure that the data output by the operator is consistent with Spark, because the final agg of the decimal sum may fallback to being executed in Spark for some reason. We need to make sure that the meaning of the intermediate data is consistent with Spark. Spark uses isEmpty and sum to distinguish many different situations for all-null input and overflow.

Reply:
> the final agg of the decimal sum may fallback to being executed in Spark

Got it. Makes sense. Perhaps, clarify this in the comments.